tcp-native/src/main/java/org/ballerinalang/stdlib/tcp/TcpClient.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.ballerinalang.stdlib.tcp;
+
+import io.ballerina.runtime.api.Future;
+import io.netty.bootstrap.Bootstrap;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.handler.timeout.IdleStateHandler;
+
+import java.net.InetSocketAddress;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link TcpClient} creates the tcp client and handles all the network operations.
+ */
+public class TcpClient {
+
+    private Channel channel;
+    private final EventLoopGroup group;
+    private final Bootstrap clientBootstrap;
+
+    public TcpClient(InetSocketAddress localAddress, InetSocketAddress remoteAddress, EventLoopGroup group,
+                     Future callback)
+            throws Exception {
+        this.group = group;
+        clientBootstrap = new Bootstrap();
+        clientBootstrap.group(group)
+                .channel(NioSocketChannel.class)
+                .handler(new ChannelInitializer<SocketChannel>() {
+                    @Override
+                    protected void initChannel(SocketChannel ch) throws Exception {
+                        ch.pipeline().addLast(Constants.CLIENT_HANDLER, new TcpClientHandler());
+                    }
+                });
+        ChannelFuture future = clientBootstrap.connect(remoteAddress, localAddress).sync()
+                .addListener((ChannelFutureListener) channelFuture -> {
+            if (channelFuture.isSuccess()) {
+                channelFuture.channel().config().setAutoRead(false);
+                channel = channelFuture.channel();
+                callback.complete(null);
+            } else {
+                callback.complete(Utils.createSocketError("Unable to connect with remote host."));
+            }
+        });
+    }
+
+    public void writeData(byte[] bytes, Future callback) throws InterruptedException {
+        if (channel.isActive()) {
+            channel.writeAndFlush(Unpooled.wrappedBuffer(bytes)).addListener((ChannelFutureListener) future -> {
+                if (future.isSuccess()) {
+                    callback.complete(null);
+                } else {
+                    callback.complete(Utils.createSocketError("Failed to send data"));
+                }
+            });
+            return;
+        }
+        callback.complete(Utils.createSocketError("Socket connection already closed."));
```

Comment:
Shall we write it as if/else blocks. It is easy to read the code that way.

Reply:
changed